identity/v2/service.go

```diff
@@ -0,0 +1,306 @@
+	ValidateAccount(ctx context.Context, accountID *types.AccountID) error
+
+	GetLastKeyByPurpose(ctx context.Context, accountID *types.AccountID, keyPurpose keystoreType.KeyPurpose) (*types.Hash, error)
+}
+
+type service struct {
+	configService        config.Service
+	centAPI              centchain.API
+	keystoreAPI          keystore.API
+	protocolIDDispatcher dispatcher.Dispatcher[protocol.ID]
+	log                  *logging.ZapEventLogger
+}
+
+func NewService(
+	configService config.Service,
+	centAPI centchain.API,
+	keystoreAPI keystore.API,
+	protocolIDDispatcher dispatcher.Dispatcher[protocol.ID],
+) Service {
+	log := logging.Logger("identity-service-v2")
+
+	return &service{
+		configService,
+		centAPI,
+		keystoreAPI,
+		protocolIDDispatcher,
+		log,
+	}
+}
+
+func (s *service) CreateIdentity(ctx context.Context, req *CreateIdentityRequest) (config.Account, error) {
+	if err := s.validateCreateIdentityRequest(ctx, req); err != nil {
+		s.log.Errorf("Invalid request: %s", err)
+
+		return nil, err
+	}
+
+	signingPublicKey, signingPrivateKey, err := generateDocumentSigningKeys()
+
+	if err != nil {
+		s.log.Errorf("Couldn't generate document signing key pair: %s", err)
+
+		return nil, ErrSigningKeyPairGeneration
+	}
+
+	acc, err := configstore.NewAccount(
+		req.Identity,
+		signingPublicKey,
+		signingPrivateKey,
+		req.WebhookURL,
+		req.PrecommitEnabled,
+	)
+
+	if err != nil {
+		s.log.Errorf("Couldn't create account: %s", err)
+
+		return nil, ErrAccountCreation
+	}
+
+	if err := s.configService.CreateAccount(acc); err != nil {
+		s.log.Errorf("Couldn't store account: %s", err)
+
+		return nil, ErrAccountStorage
+	}
+
+	protocolID := p2pcommon.ProtocolForIdentity(acc.GetIdentity())
+
+	err = s.protocolIDDispatcher.Dispatch(ctx, protocolID)
+
+	if err != nil {
+		s.log.Errorf("Couldn't dispatch protocol ID: %s", err)
+
+		return nil, ErrProtocolIDDispatch
+	}
+
+	return acc, nil
+}
+
+func generateDocumentSigningKeys() (libp2pcrypto.PubKey, libp2pcrypto.PrivKey, error) {
+	signingPublicKey, signingPrivateKey, err := ed25519.GenerateSigningKeyPair()
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	privateKey, err := libp2pcrypto.UnmarshalEd25519PrivateKey(signingPrivateKey)
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	publicKey, err := libp2pcrypto.UnmarshalEd25519PublicKey(signingPublicKey)
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return publicKey, privateKey, nil
+}
+
+func (s *service) validateCreateIdentityRequest(ctx context.Context, req *CreateIdentityRequest) error {
+	accID, err := types.NewAccountID(req.Identity[:])
+
+	if err != nil {
+		s.log.Errorf("Couldn't create account ID: %s", err)
+
+		return ErrAccountIDCreation
+	}
+
+	if err := s.ValidateAccount(ctx, accID); err != nil {
+		s.log.Errorf("Invalid identity - %s: %s", accID.ToHexString(), err)
+
+		return ErrInvalidAccount
+	}
+
+	if _, err := url.ParseRequestURI(req.WebhookURL); err != nil {
```

Comment:
Since the webhook is optional, I would only fail here if there is a WebhookURL provided and the format is incorrect.

Reply:
OK.